packages/react-router-dev/vite/plugin.ts

```diff
@@ -916,31 +933,59 @@ export const reactRouterVitePlugin: ReactRouterVitePlugin = () => {
       },
       routes,
     };
+
+    currentReactRouterManifestForDev = reactRouterManifestForDev;
+
+    return reactRouterManifestForDev;
   };
 
-  // We use a separate environment for loading the server manifest and inlined
-  // CSS during development. This is because "ssrLoadModule" isn't available if
-  // the "ssr" environment has been defined by another plugin (e.g.
-  // vite-plugin-cloudflare) as a custom Vite.DevEnvironment rather than a
-  // Vite.RunnableDevEnvironment:
-  // https://vite.dev/guide/api-environment-frameworks.html#runtime-agnostic-ssr
-  const HELPER_ENVIRONMENT_NAME = "__react_router_helper__";
+  const loadCssContents: LoadCssContents = async (viteDevServer, dep) => {
+    invariant(
+      viteCommand === "serve",
+      "loadCssContents is only available in dev mode"
+    );
 
-  const loadModule: LoadModule = (viteDevServer, url) => {
-    if (ctx.reactRouterConfig.future.unstable_viteEnvironmentApi) {
-      const vite = getVite();
-      const helperEnvironment =
-        viteDevServer.environments[HELPER_ENVIRONMENT_NAME];
+    if (dep.file && isCssModulesFile(dep.file)) {
+      return cssModulesManifest[dep.file];
+    }
 
+    const vite = getVite();
+    const viteMajor = parseInt(vite.version.split(".")[0], 10);
+
+    const url =
+      viteMajor >= 6
+        ? // We need the ?inline query in Vite v6 when loading CSS in SSR
+          // since it does not expose the default export for CSS in a
+          // server environment. This is to align with non-SSR
+          // environments. For backwards compatibility with v5 we keep
+          // using the URL without ?inline query because the HMR code was
+          // relying on the implicit SSR-client module graph relationship.
+          injectQuery(dep.url, "inline")
+        : dep.url;
+
+    let cssMod: unknown;
+    if (ctx.reactRouterConfig.future.unstable_viteEnvironmentApi) {
+      const cssDevHelperEnvironment =
+        viteDevServer.environments[CSS_DEV_HELPER_ENVIRONMENT_NAME];
       invariant(
-        helperEnvironment && vite.isRunnableDevEnvironment(helperEnvironment),
-        "Missing helper environment"
+        cssDevHelperEnvironment &&
+          vite.isRunnableDevEnvironment(cssDevHelperEnvironment),
+        "Missing CSS dev helper environment"
       );
-
-      return helperEnvironment.runner.import(url);
+      cssMod = await cssDevHelperEnvironment.runner.import(url);
+    } else {
+      cssMod = await viteDevServer.ssrLoadModule(url);
     }
 
-    return viteDevServer.ssrLoadModule(url);
+    invariant(
+      typeof cssMod === "object" &&
+        cssMod !== null &&
+        "default" in cssMod &&
+        typeof cssMod.default === "string",
+      `Failed to load CSS for ${dep.url}`
+    );
```

Comment:
The only consumer of this code threw an error if the result was `undefined`, so I've inlined that logic into this function now instead.